skglm/solvers/prox_newton_solver.py

```diff
@@ -0,0 +1,312 @@
+            else:
+                _prox_newton_iter(
+                    X, Xw, w, y, penalty, ws, min_pn_cd_epochs, _max_pn_cd_epochs,
+                    max_backtrack, pn_tol)
+
+            # 3) do Anderson acceleration on smaller problem
+            w_acc, Xw_acc = accelerator.extrapolate(w, Xw)
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+            p_obj_acc = datafit.value(y, w_acc, Xw_acc) + penalty.value(w_acc)
+
+            if p_obj_acc < p_obj:
+                w[:] = w_acc
+                Xw[:] = Xw_acc
+
+            if epoch % 5 == 0:  # check every 5 epochs, PN epochs are expensive
+                p_obj = datafit.value(y, w[ws], Xw) + penalty.value(w)
+
+                if is_sparse:
+                    grad = construct_grad_sparse(
+                        X.data, X.indptr, X.indices, y, w, Xw, datafit, ws)
+                else:
+                    grad = construct_grad(X, y, w, Xw, datafit, ws)
+                if ws_strategy == "subdiff":
+                    opt_ws = penalty.subdiff_distance(w, grad, ws)
+                elif ws_strategy == "fixpoint":
+                    opt_ws = dist_fix_point(w, grad, datafit, penalty, ws)
+
+                stop_crit_in = np.max(opt_ws)
+                if max(verbose - 1, 0):
+                    print(f"Epoch {epoch + 1}, objective {p_obj:.10f}, "
+                          f"stopping crit {stop_crit_in:.2e}")
+                if ws_size == n_features:
+                    if stop_crit_in <= tol:
+                        break
+                else:
+                    if stop_crit_in < 0.3 * stop_crit:
+                        if max(verbose - 1, 0):
+                            print("Early exit")
+                        break
+        obj_out.append(p_obj)
+    return w, np.array(obj_out), stop_crit
+
+
+@njit
+def _prox_newton_iter(
+    X, Xw, w, y, penalty, ws, min_cd_epochs, max_cd_epochs, max_backtrack, tol
+):
+    n_samples, ws_size = X.shape[0], ws.shape[0]
+
+    hessian_diag = np.zeros(n_samples)  # hessian = X^T D X, with D = diag(f_i'')
+    grad_datafit = np.zeros(n_samples)  # gradient of F(Xw)
+
+    lc = np.zeros(ws_size)  # weighted Lipschitz constants
+    bias = np.zeros(ws_size)
+
+    for i in range(n_samples):
+        prob = 1. / (1. + np.exp(y[i] * Xw[i]))
+        # this supposes that the datafit is scaled by n_samples
+        hessian_diag[i] = prob * (1. - prob) / n_samples
+        grad_datafit[i] = -y[i] * prob / n_samples
+
+    for idx, j in enumerate(ws):
+        lc[idx] = weighted_dot(X, Xw, hessian_diag, j, ignore_b=True)
+        bias[idx] = X[:, j] @ grad_datafit
+
+    delta_w, X_delta_w = _newton_cd(
+        X, w, ws, hessian_diag, bias, lc, penalty, min_cd_epochs, max_cd_epochs, tol)
+    step_size = _backtrack_line_search(
+        w, Xw, delta_w, X_delta_w, ws, y, penalty, max_backtrack)
+
+    for idx, j in enumerate(ws):
+        w[j] += step_size * delta_w[idx]
+    Xw += step_size * X_delta_w
+
+
+@njit
```

Comment:
As a first version I'd support only dense X to split the work in easier pieces @Badr-MOUFAD @PABannier